### PR TITLE
find-esp: rework verify_fsroot_dir()  a bit

### PR DESCRIPTION
**Pull request by poettering**

### CODE REVIEW

File: src/basic/stat-util.c [hunk 1]
```diff
@@ -415,3 +415,30 @@ int statx_fallback(int dfd, const char *path, int flags, unsigned mask, struct s
...
+        if (fstat(fd, &sta) < 0)
+                return -errno;
+
+        if (!S_ISDIR(sta.st_mode)) /* If it's not a dir, it's definitely not the root dir */
+                return false;
+
+        if (fstatat(fd, "..", &stb, 0) < 0) {
+
+                if (ERRNO_IS_PRIVILEGE(errno)) /* If the parent dir has different access controls, then it's not the same inode */
+                        return 0;
```
Comment (yuwata) [inline src/basic/stat-util.c]: `0` -> `false` is better, for consistency to other places.

**Review by yuwata — APPROVED** [review]

File: src/shared/find-esp.c [hunk 2]
```diff
@@ -283,21 +285,24 @@ static int verify_fsroot_dir(
                         if (r < 0)
                                 return log_error_errno(r, "Failed to extract parent path from '%s': %m", path);
 
-                        r = RET_NERRNO(stat(parent, &st2));
-                }
+                        r = get_block_device(parent, &devt_b);
```
Comment (yuwata) [inline src/shared/find-esp.c]: BTW, in what situation this fallback logic works??

**Comment by poettering** [comment]
hmm CI looks suspicious

File: src/shared/find-esp.c (hunk 2 above)
Comment (poettering) [reply src/shared/find-esp.c]: i have a testing setup here where /boot/ and /boot/efi are set up the way fedora does by default, i.e. /boot is a separate partition and /boot/efi is the ESP. If you are unprivileged you can enumerate the contents of /boot, but /boot/efi is made inaccessible via the dmask= mount option: you may stat() the /boot/efi inode itself but may not enumerate or access anything inside it. This means we /boot/ is accessible, but /boot/efi/.. is not...

hence this annyoing fallback.

I'll add a longer comment.

Comment (yuwata) [reply src/shared/find-esp.c]: Interesting, thank you for the explanation.

**Comment by poettering** [comment]
fixed the one thing @yuwata found, plus fixed one more thing. Let's see if this passes CI now.

**Comment by yuwata** [comment]
Hm,,, still all mkosi CIs are red...

**Comment by poettering** [comment]
OK, I substantially reworked this now around statx(). Dropping green label. I hope it passes CI this time.

PTAL!

File: src/basic/stat-util.c [hunk 3]
```diff
@@ -360,6 +360,19 @@ bool stat_inode_unmodified(const struct stat *a, const struct stat *b) {
                 (!(S_ISCHR(a->st_mode) || S_ISBLK(a->st_mode)) || a->st_rdev == b->st_rdev); /* if device node, also compare major/minor, because we can */
```
Comment (yuwata) [inline src/basic/stat-util.c:360]: commit message: tham -> than

File: src/shared/find-esp.c [hunk 4]
```diff
@@ -8,11 +8,13 @@
```
Comment (yuwata) [inline src/shared/find-esp.c:8]: commit message: "wants if if then wants"??

File: src/shared/find-esp.c [hunk 5]
```diff
@@ -229,75 +231,127 @@ static int verify_esp_udev(
         return 0;
 }
 
+static int statx_mount_same(const struct statx *a, const struct statx *b) {
```
Comment (yuwata) [inline src/shared/find-esp.c]: Why not to move this to stat-util.c?

File: src/shared/find-esp.c [hunk 6]
```diff
@@ -229,75 +231,127 @@ static int verify_esp_udev(
...
+
+                        if (path_equal(path, "/"))
+                                goto success;
 
                         r = path_extract_directory(path, &parent);
                         if (r < 0)
                                 return log_error_errno(r, "Failed to extract parent path from '%s': %m", path);
 
-                        r = RET_NERRNO(stat(parent, &st2));
+                        r = statx_fallback(AT_FDCWD, parent, AT_SYMLINK_NOFOLLOW, STATX_TYPE, &sxb);
```
Comment (yuwata) [inline src/shared/find-esp.c]: Why only STATX_TYPE here?

File: src/shared/find-esp.c [hunk 7]
```diff
@@ -229,75 +231,127 @@ static int verify_esp_udev(
...
 
-        t2 = strjoina(path, "/..");
-        if (stat(t2, &st2) < 0) {
-                if (errno != EACCES)
-                        r = -errno;
-                else {
+        /* Now let's look at the parent */
+        r = statx_fallback(fd, "..", 0, STATX_TYPE|STATX_INO|STATX_MNT_ID, &sxb);
+        if (r < 0) {
+                if (ERRNO_IS_PRIVILEGE(r)) {
```
Comment (yuwata) [inline src/shared/find-esp.c]: You can merge the conditions, and one level of indentation can be reduced.

File: src/shared/find-esp.c [hunk 8]
```diff
@@ -229,75 +231,127 @@ static int verify_esp_udev(
...
+                /* The statx() above we can execute on an O_PATH fd. But the btrfs ioctl we cannot. Hence
+                 * acquire a "real" fd first, without the O_PATH flag. */
+
+                real_fd = fd_reopen(fd, O_DIRECTORY|O_CLOEXEC);
+                if (real_fd < 0)
+                        return real_fd;
+
+                r = btrfs_get_block_device_fd(real_fd, ret_dev);
+                if (r < 0)
+                        return r;
```
Comment (yuwata) [inline src/shared/find-esp.c]: Not a matter, but I'd like `return btrfs_get_block_device_fd()`.

**Comment by yuwata** [comment]
CIs are still red:
```
../../src/systemd/src/shared/find-esp.c:240:27: error: no member named 'stx_mnt_id' in 'struct statx'
                return a->stx_mnt_id == b->stx_mnt_id;
                       ~  ^
../../src/systemd/src/shared/find-esp.c:240:44: error: no member named 'stx_mnt_id' in 'struct statx'
                return a->stx_mnt_id == b->stx_mnt_id;
                                        ~  ^
2 errors generated.
```

It seems that find-esp.c needs to use `STRUCT_NEW_STATX_DEFINE` instead of `STRUCT_STATX_DEFINE`.

File: src/shared/find-esp.c (hunk 6 above)
Comment (poettering) [reply src/shared/find-esp.c]: bug!

**Comment by poettering** [comment]
Force pushed a new version, with all points addressed, and hopefully CI should be happy now too.

**Review by yuwata — APPROVED** [review]
LGTM.

**Pull request merged**